[EC-281] Add de/serialization methods to CipherView objects

eliykat opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->
Add de/serialization methods to `CipherView` and its nested objects. This is to support communicating with web workers, which can only pass serialized data, not object references.

This enables #3043, which alleviates performance issues in #1597.

## Code changes

The approach here is:
* serialization:
  * `JSON.stringify` will call the `toJSON()` method on the object (and recursively on any nested object) if it has that method, then serializes that return value. Use this where required to handle special cases (commented in the code)
  * Otherwise, rely on the documented `JSON.stringify` behaviour wherever possible - i.e. let it handle all enumerated own properties, let it handle date conversions. No need to reinvent the wheel.

* deserialization:
  * add a `fromJSON(obj)` static method on all classes. This is not treated specially by `JSON.parse`, but it's helpful to follow the same convention.
  * ~this method generally uses a new helper method, `Utils.copyToNewObject`, to create and hydrate a new object with primitive values~ (EDIT: removed due to feedback)
  * this method also handles any nested objects, like instantiating `Date`s or our own nested objects.

I've also added tests.

Please also closely review my changes to `symmetricCryptoKey`, as they adjust recent changes made there.

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

## Before you submit

<!-- (mark with an `X`) -->

```
- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [x] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

eliykat commented:
I've decided the tests probably should be broken up, but would appreciate feedback in the meantime.

eliykat commented on libs/common/src/models/view/passwordHistoryView.ts:
```diff
@@ -13,4 +15,18 @@ export class PasswordHistoryView implements View {
   }
+
+  static fromJSON(obj: any): PasswordHistoryView {
+    const ph = Utils.copyToNewObject(
+      obj,
+      {
+        password: null,
+      },
+      PasswordHistoryView
+    );
+
+    ph.lastUsedDate = obj.lastUsedDate == null ? null : new Date(obj.lastUsedDate);
+
+    return ph;
```
Fair enough - I wasn't sure how far to go in reducing boilerplate, but sometimes boring & more verbose is best. I'll look into it.

eliykat commented:
I've addressed feedback and restructured tests. Also added typing in the form of `Partial<ViewClass>` instead of `any`.

Hinton commented:
I think the `Partial` could be slightly misleading. I use it when only some fields are set. I don't expect the type to differ, just to potentially be null.

Hinton commented on libs/common/src/models/view/attachmentView.ts:
```diff
@@ -32,4 +32,17 @@ export class AttachmentView implements View {
     return 0;
   }
+
+  static fromJSON(obj: Partial<AttachmentView>): AttachmentView {
```
I would like to have an interface for the `fromJson` methods. `Deserializeable` or something.

eliykat replied:
Unfortunately there doesn't appear to be any way to enforce an interface for a static method (or have `abstract` static methods). There are lots of long discussions about it, see https://github.com/microsoft/TypeScript/issues/13462 (starting in 2017), which was moved to https://github.com/microsoft/TypeScript/issues/14600, which was split to https://github.com/microsoft/TypeScript/issues/34516 and https://github.com/microsoft/TypeScript/issues/33892, with still no clear outcome or recommended workaround.

eliykat replied:
Unfortunately there doesn't appear to be any way to enforce an interface for a static method (or have `abstract` static methods). There are lots of long discussions about it, see https://github.com/microsoft/TypeScript/issues/14600, which was split to https://github.com/microsoft/TypeScript/issues/34516 and https://github.com/microsoft/TypeScript/issues/33892. But no clear outcome or good, recommended workaround that I can see.

Hinton commented on libs/common/src/models/view/cardView.ts:
```diff
@@ -79,4 +79,28 @@ export class CardView extends ItemView {
     return year.length === 2 ? "20" + year : year;
   }
+
+  toJSON(): Partial<CardView> {
```
Same here, we should have an interface for `toJSON`.

eliykat replied:
Classes only implement `toJSON` if the default serialization behaviour isn't suitable. So not all serializable classes implement it, at best it's an optional property, which is less useful.

I'm open to just having a `serializable` interface just to indicate that "we've looked at this and it can be safely de/serialized". But I don't think it'll do much to actually enforce an interface.

eliykat commented:
> I think the `Partial` could be slightly misleading. I use it when only some fields are set. I don't expect the type to differ, just to potentially be null.

TS Handbook on the [Partial utility type](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype) says:

> Constructs a type with all properties of Type set to optional. This utility will return a type that represents all subsets of a given type.

So it's not really for nullable fields (i.e. where the value might be `null`), it's for optional fields (i.e. where the field might not exist on the object). And so by definition it'll be a different type.
 
> What do you think about doing something like this.
> 
> ```ts
> type JsonSerialized<T extends object> = {
>     [K in keyof T]: T[K] extends Date ? string
>         : T[K]
> }
> 
> type JsonSerializedFolderView = JsonSerialized<FolderView>;
> ```

This is what `Partial` does:

![Screen Shot 2022-07-13 at 9 12 54 am](https://user-images.githubusercontent.com/31796059/178614671-95c1bc68-5fb5-40de-882e-80e62249735a.png)

eliykat commented on libs/common/src/types/serializationTypes.ts:
```diff
@@ -0,0 +1,28 @@
+ * Represents an object that can be serialized using JSON.stringify() without data loss
+ * Usually the type returned by a toJSON() method
+ */
+export type SerializableObject<T extends object> = {
```
Explanation:
* if T's property can already be serialized, then this property will be the same type (no change)
* otherwise, we don't know what type it needs to be transformed to, so make it `any[]` if an array or `any` otherwise. (Note this will not necessarily be a string, because it's not serialized yet)

eliykat commented on libs/common/src/types/serializationTypes.ts:
```diff
@@ -0,0 +1,28 @@
+ * Represents an object returned by JSON.parse() before it's properly rehydrated by a ctor or static method
+ * Usually the type given to fromJSON()
+ */
+export type ParsedObject<T extends object> = {
```
Explanation:
* if T's property can already be de/serialized, then this property will be the same type (no change), JSON.parse() will handle it for us
* if T's property is a Date, then we know that'll be stored as a string and will remain a string after `JSON.parse()` is called
* otherwise, we don't know what it is, so either `any[]` or `any`.

Hinton commented:
I'm happy with this change. My objection to `Partial` was that the types could be different, i.e. a date was actually a string. This solves that 👍.

MGibson1 commented on libs/common/src/models/domain/symmetricCryptoKey.ts:
```diff
@@ -55,21 +56,17 @@ export class SymmetricCryptoKey {
     }
 
-    return jsonResult;
+    const arrayBuffer = Utils.fromB64ToArray(obj.keyB64).buffer;
```
I had bugs when trying to recreate from just the full key. You're 100% right it should work, but it didn't for me 🤷 .

eliykat replied:
I see you're using `Object.create`. I haven't used it before, but Google says it doesn't run constructors. Maybe that's the problem, the `withPrototype` decorator wasn't calling the ctor to parse the `key` buffer?

I don't think this PR will have the same problem, but can you give me some pointers on how I can do a quick regression test on this?

MGibson1 replied:
Object.create basically smashes a prototype onto your object. It's a sledge hammer.

 The real issue I saw was when a `converter` was supplied as in https://github.com/bitwarden/clients/blob/master/libs/common/src/services/state.service.ts#L486.
 
 If I initialized from just the key I saw encryption failures when unlocking in manifest v3 only. You can test with `npm run build:watch:mv3` and poking around the browser extension.

MGibson1 reviewed (commented):
I would like to see a class that these methods are extending. Then we can require that class on anything that is stored by stateService

MGibson1 commented:
I held this for `rc` cuts and the discuss how serialization from json will practically work.

eliykat commented:
@MGibson1 I'd also like this, my 

> Since we can't have statics in an interface, here's my thinking:
> 
> ```ts
> abstract class Storable {
>   abstract toJSON(): string;

This method only needs to be implemented if the default behaviour of [JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) is not suitable (i.e. complex objects). Many classes in this PR do not implement this method because it's not required, but this doesn't mean they're not serializable.

> 
>   /** 
>    * Map of known storable types to their json constructor functions.
>    */
>   private static knownTypes = new Map<new (...args: any) => any, (json: string) => any>([
> 
>   ]);
> 
>   static fromJSON<T>(json: string, ctor: new () => T): T {
>     if (this.knownTypes.has(ctor)) {
>       return this.knownTypes.get(ctor)(json);
>     } else {
>       throw new Error(`Unknown storable type: ${ctor.name}`);
>     }
>   }
> }
> ```

I don't understand how this improves things, instead of calling `MyViewClass.toJSON(obj)`, we're calling `Storable.fromJSON(obj, MyViewClass.fromJSON)`? (or calling the same staic method from the child view class?) How does this enforce an interface (or similar)?

> Then we would stick the known Storable types in there to call from whatever is doing the initializing of our JSON data, probably the `withPrototype` decorator.

I like the idea of using a decorator to do this, it's pretty slick. I think you could still do this without the Storable map though. (just pass in `MyViewClass.fromJSON` as the arg)

eliykat commented:
Relatedly - just thinking aloud - I don't even know how we'd enforce a `Storable` interface in the current `StateService`, every setter method expects a specific type. Is there any way to generally enforce a rule that any object being stored to disk must be `Storable`? Maybe a runtime `instanceof` check?

MGibson1 commented:
> This method only needs to be implemented if the default behaviour of [JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) is not suitable (i.e. complex objects). Many classes in this PR do not implement this method because it's not required, but this doesn't mean they're not serializable.

I think we're getting our wires crossed a little. The interface implementation is for `toJSON` where it's actually useful both to provide a way to serialize otherwise un-serializable data and to remind users of storage that they need to think about serialization, not just magically persisting program state.

The `fromJSON` doesn't require implementation anywhere (mostly because I can't see a way to do it -- I'd love for it to be required too).

The improvement I suggest for registering these deserializers is just that the `withPrototype` decorator would be able to take exactly one argument, the constructor of the object. It actually already optionally accepts a `fromJSON` converter ([here](https://github.com/bitwarden/clients/blob/master/libs/common/src/services/state.service.ts#L2741-L2744)). I'm fine with continuing this pattern if you prefer. I was mostly spitballing to try and centralize json serialization concerns.

>I don't understand how this improves things, instead of calling MyViewClass.toJSON(obj), we're calling Storable.fromJSON(obj, MyViewClass.fromJSON)? (or calling the same staic method from the child view class?) How does this enforce an interface (or similar)?

You wouldn't call this method with `Storable.fromJSON(obj, MyViewClass.fromJSON)`, it would just be `Storable.fromJSON(obj, MyViewClass)`. My thinking was that we could pass an object in this method along with its constructor, it would search for known types and initialize them for us. However, I've just done some testing. Turns out you _can_ just call static methods on a class from the constructor. I thought they were stored elsewhere for some reason. It's still probably preferable to pass in an initializer like you suggest. I just spent the last hour trying to be cleverer and it's best to be explicit here.

eliykat commented:
@MGibson1 Thanks for the chat the other day, very helpful.

I've done the following:
* added `Storable` abstract class - to make some of our patterns more explicit
* added `toJSON` methods that just return `this` where required
* added tests for default toJSON behaviour
* slightly tweaked the types on `withPrototype` decorator functions. Technically the converter function does not take `T` and return `T`, it could take a `ParsedObject<T>` as well (that's why we need the converter function, to instantiate `T`). For now I've just fudged it with `any` as I'm not sufficiently certain about the exact type it could be receiving.

MGibson1 commented on libs/common/src/models/storable.ts:
```diff
@@ -0,0 +1,43 @@
+export type ToJsonObject<T extends object> = {
+  [K in keyof T]?: T[K] extends SerializablePrimitives | SerializablePrimitives[]
+    ? T[K]
+    : T[K] extends Array<any>
+    ? any[]
+    : any;
```
Can we do better than these `any`s? they need to extend `Storable`, but I don't know how to get a generic type reference to a second-order generic...

MGibson1 replied:
It looks like you can recurse types, so this could become:

```typescript
export type ToJsonObject<T> = {
  [K in keyof T]?: T[K] extends SerializablePrimitives | SerializablePrimitives[]
    ? T[K]
    : T[K] extends Array<ToJsonObject<any>>
    ? ToJsonObject<any>[]
    : ToJsonObject<T[K]>;
};
```

but I'm struggling to get rid of the `any` in the array type check...

MGibson1 replied:
I keep messing around with this and eventually came up with this. who knows if it works 😅 

```typescript
export type ToJsonObject<T extends object> = {
  [K in keyof T]?: T[K] extends SerializablePrimitives | SerializablePrimitives[]
    ? T[K]
    : T[K] extends ArrayOfSerializablePrimitives
    ? SerializablePrimitives[]
    : T[K] extends Array<any>
    ? T[K][0] extends object
      ? ToJsonObject<T[K][0]>[]
      : never
    : T[K] extends object
    ? ToJsonObject<T[K]>
    : never;
};

type ArrayOfSerializablePrimitives = Array<SerializablePrimitives>;
```

I get the feeling that these helper types will simplify this type inclusion logic immensely. I'd love to spend some time going down this rabbit hole, but I'm short atm. Perhaps you can spend a bit on this?

MGibson1 replied:
oh also, here's the horrible thing I was trying to make: https://github.com/sindresorhus/type-fest/blob/main/source/jsonify.d.ts#L65-L91

Hinton replied:
@MGibson1 should we add type-fest as a dependency? Seems they have some useful helpers.

eliykat replied:
Added type-fest as a dependency and re-did types, see latest comment. Please create new comment thread if required

MGibson1 replied:
yup that's the best solution for sure

MGibson1 commented on libs/common/src/models/storable.ts:
```diff
@@ -0,0 +1,43 @@
+ * You must also implement a static fromJSON method - see the example interface below
+ * (unfortunately we can't enforce static methods from an abstract class)
+ */
+export abstract class Storable<T extends object> {
```
I don't think this needs to be a class for any reason, right?

If it's an interface, you no longer have the inheritance chain for `__View` -> `ItemView<T>` -> `Storable<T>`, you can just `implement` directly on the `__View`s.

eliykat replied:
Yes that's much better, I didn't like that chain of inheritance but wasn't sure what do about it.

MGibson1 commented on libs/common/src/models/view/itemView.ts:
```diff
@@ -1,8 +1,9 @@
 import { View } from "./view";
 
-export abstract class ItemView implements View {
+export abstract class ItemView<T extends object> extends Storable<T> implements View {
```
The fact that the type is not used in this abstract indicates that it probably shouldn't extent `Storable` itself. I know you did this because you can only inherit from one class, but `Storable` should be changed to an interface, then you're good.

MGibson1 commented on libs/common/src/models/storable.ts:
```diff
@@ -0,0 +1,43 @@
+ * An object returned by JSON.parse() before it's reinitialized as a class instance by fromJSON
+ */
+export type ParsedObject<T extends object> = {
+  [K in keyof T]?: T[K] extends DeserializablePrimitives | DeserializablePrimitives[]
+    ? T[K]
+    : T[K] extends Date
+    ? string
+    : T[K] extends Array<any>
+    ? any[]
+    : any;
+};
```
Should be possible to do somthing similar here as for `toJsonObject`

MGibson1 commented on libs/common/src/models/view/cipherView.ts:
```diff
@@ -131,4 +135,93 @@ export class CipherView implements View {
+    return result;
+  }
+
+  static fromJSON(obj: ParsedObject<CipherView>): CipherView {
```
Each of these comment sections can be split up into private sub-methods to better logically encapsulate work

eliykat replied:
Done. Left the simple property assignments in `fromJSON`, but extracted dates and nested view objects to their own private methods.

MGibson1 reviewed (changes requested):
This largely looks really good. I think there's a little bit of tightening up of the definitions of `Storable`, then this is ready to go.

I also hate to spring this now, but pehaps all this work is indicative that we should think about a package (https://www.npmjs.com/package/@upe/typedjson perhaps?)

I also found this type definitions package that has this lovely little guy in there: https://github.com/sindresorhus/type-fest/blob/main/source/basic.d.ts#L20-L22

eliykat commented:
> I also hate to spring this now, but pehaps all this work is indicative that we should think about a package (https://www.npmjs.com/package/@upe/typedjson perhaps?)
> 
> I also found this type definitions package that has this lovely little guy in there: https://github.com/sindresorhus/type-fest/blob/main/source/basic.d.ts#L20-L22

I think it's a good idea to bring in external type definitions to avoid having to roll our own. Both our attempts are pretty ugly 😂 

On the topic of using an external library - I'll evaluate this next week and get back to you. I also found https://github.com/shaochuancs/esserializer, which looks a bit more recent. Feels a bit silly to pivot at this late stage, but if it gets us a more maintainable/robust solution with minimal effort, then we may as well do it now.

eliykat commented:
Okay @MGibson1 I've spent some more time on this, here's where I've landed.

I tried out esserializer to handle this, but I hit a few errors trying to integrate it with our repo, which prevented me from trying it properly and which made me wary of relying on it. (Even though I do think it's pretty neat.) It has some nice advanced options for handling complex cases, but that made me realise that it's probably just moving the work around - instead of writing our own methods, we're trying to fit our case into the external library's api. I decided that it's better to keep the current approach, which anyone can look at and understand almost immediately. We can rely on the `Storable` interface + types (+ unit tests) to keep it maintainable.

Same with the typedjson package you mentioned, and I like scattering decorators through our classes less.

I did add `type-fest` as a dependency, and adapt its [Jsonify type](https://github.com/sindresorhus/type-fest/blob/main/source/jsonify.d.ts#L64) for our use. I think the typing is much more solid and readable now, check it out. (moved to the `types/json.types.ts` file)

MGibson1 commented on libs/common/src/types/json.types.ts:
```diff
@@ -0,0 +1,89 @@
+
+/**
+ * Represents a type that can be safely serialized using JSON.stringify without data loss.
+ * Adapted from type-fest's jsonify type for our own needs
```
can you expand in this comment what was added in your tweaks? This is fairly impenetrable.

eliykat replied:
I've scrapped this in favour of `Jsonify` as discussed

MGibson1 reviewed (changes requested):
regarding the `jsonify` type.

In practice this type is accomplishing a lot of the same things that the `storable` interface does. The only real "benefit" (in quotes because it's dubious at best) is the opportunity to provide tsdocs warning you to implement a fromJSON as well.

I'm no longer convinced an interface is the better way to do this and instead just have our consumers of this storable expect a `jsonify`-type object. Then, where `fromJSON` is necessary, we can again require a callback of type `(obj: jsonify<T>) => T`. Should do a good job guarding our methods from objects that will cause issues.

eliykat commented on libs/common/src/models/view/cardView.ts:
```diff
@@ -79,4 +78,28 @@ export class CardView extends ItemView {
+    };
+  }
+
+  static fromJSON(obj: FromJson<CardView>): CardView {
+    const view = new CardView();
+    view.cardholderName = obj.cardholderName;
+    view.brand = obj.brand;
+    view.number = obj.number;
+    view.expMonth = obj.expMonth;
+    view.expYear = obj.expYear;
+    view.code = obj.code;
+
+    return view;
+  }
```
very nice one-liner, didn't realise that `Object.assign` returned the target value!

eliykat commented on libs/common/src/models/view/cipherView.ts:
```diff
@@ -131,4 +134,98 @@ export class CipherView implements View {
   }
+
+  toJSON() {
+    const result: ToJson<CipherView> = {
```
You're right, now that I look at it again we don't need a `toJSON` method on this class at all.

eliykat commented on libs/common/src/models/view/identityView.ts:
```diff
@@ -139,4 +142,52 @@ export class IdentityView extends ItemView {
   }
+
+  toJSON() {
+    // Needed to serialize getters which are not included by JSON.stringify
```
Agreed that we'll use private properties instead, to capture & replicate internal state, rather than using the public interface.

eliykat commented on libs/common/src/models/view/loginUriView.ts:
```diff
@@ -124,4 +127,19 @@ export class LoginUriView implements View {
   }
+
+  toJSON() {
+    // Needed to serialize getters which are not included by JSON.stringify
```
Have removed this pattern as discussed, we'll stringify & assign back in the internal state

eliykat commented:
@MGibson1 I've actioned your feedback:
* removed Storage interface and custom json types
* use type-fest's `Jsonify` instead
* serialize & restore internal state instead of using public setters & getters
* use `Object.assign` for most property assignments
* removed `toJSON` tests where we're just relying on JS behaviour

The result of this is drastically cutting down the overall LOC which is nice. We have almost no `toJSON` methods anymore.

Note that I'm using `Partial<Jsonify<T>>` as the `toJSON` parameter, because:
* `Jsonify<T>` includes getters as required properties in the type, when these are not always required to recreate the object. (e.g. `FieldView.maskedValue`) I noticed this because my tests started throwing compile-time errors.
* in principle, you do not need to supply a value for every property in order to create the object.

Question: do you think the remaining tests have any value? The logic is so simple that it's almost the same in both the prod and the test code.

MGibson1 commented on libs/common/spec/models/view/cardView.spec.ts:
```diff
@@ -1,30 +1,20 @@
+    const expected = Object.assign(new CardView(), testValues);
     expect(actual).toEqual(expected);
+    expect(actual).toBeInstanceOf(CardView);
   });
```
I don't think this particular test is useful because it's fully covered by the type system.

We should only need these tests if we have to validate that something unprotected happened. For example, `Object.assign` doesn't honor types, so `AttachmentView` needs to test its `fromJSON` method to ensure the symmetric key is properly deserialized into a real object.

eliykat replied:
I agree, they were more relevant when we were individually assigning properties. I'll rewrite to just test "interesting" things like object instantiation, not the assignment of primitive properties.

MGibson1 commented on libs/common/spec/models/view/fieldView.spec.ts:
```diff
@@ -3,31 +3,20 @@ import { LoginLinkedId } from "@bitwarden/common/enums/linkedIdType";
+    const expected = Object.assign(new FieldView(), testValues);
     expect(actual).toEqual(expected);
     expect(actual).toBeInstanceOf(FieldView);
   });
```
This test is also unnecessary -- fully protected by type validation

MGibson1 commented on libs/common/spec/models/view/identityView.spec.ts:
```diff
@@ -1,42 +1,31 @@
+    const expected = Object.assign(new IdentityView(), testValues);
     expect(actual).toEqual(expected);
     expect(actual).toBeInstanceOf(IdentityView);
   });
```
This test is also unnecessary -- fully protected by type validation

MGibson1 commented on libs/common/spec/models/view/loginUriView.spec.ts:
```diff
@@ -1,27 +1,16 @@
+    const expected = Object.assign(new LoginUriView(), testValues);
     expect(actual).toEqual(expected);
     expect(actual).toBeInstanceOf(LoginUriView);
   });
```
This test is also unnecessary -- fully protected by type validation

MGibson1 commented on libs/common/spec/models/view/passwordHistoryView.spec.ts:
```diff
@@ -1,27 +1,16 @@
+    const expected = Object.assign(new PasswordHistoryView(), testValues);
     expect(actual).toEqual(expected);
     expect(actual).toBeInstanceOf(PasswordHistoryView);
   });
```
This should test date instatiation, the rest of the sut method is pretty uninteresting.

MGibson1 commented on libs/common/spec/models/view/secureNoteView.spec.ts:
```diff
@@ -2,25 +2,14 @@ import { SecureNoteType } from "@bitwarden/common/enums/secureNoteType";
+    const expected = Object.assign(new SecureNoteView(), testValues);
     expect(actual).toEqual(expected);
     expect(actual).toBeInstanceOf(SecureNoteView);
   });
```
This test is also unnecessary -- fully protected by type validation

MGibson1 reviewed (dismissed):
A few suggestions, but nothing game-changing. Up to you if you want to do another round or not.

One thing I'm noting is that there are no protections on the `fromJSON` method doing a good job diving through its property tree and appropriately calling those `fromJSON` methods. We'll need a good hint comment on the consumers of these initializers.

MGibson1 reviewed (approved):
re-up approval dismissed due to merge commit

eliykat commented on libs/common/spec/models/view/attachmentView.spec.ts:
```diff
@@ -4,38 +4,23 @@ import { AttachmentView } from "@bitwarden/common/models/view/attachmentView";
-    const expected = new AttachmentView();
-    Object.assign(expected, testValues, { key: "encKeyfromJSON" });
-
+    const expected = Object.assign(new AttachmentView(), testValues, { key: "encKeyfromJSON" });
```
You're right that it could be defeated with a constant return value, but that seems really unlikely to happen in practice. I prefer testing this way because it tests that the value is actually assigned to the object property, which isn't tested by just asserting the call. And in principle, I also think it focuses more on the output expected of the SUT, rather than its internal calls (even though both approaches do have to deal with that boundary).

If you wanted total coverage, you'd both assert the call and mock the return value I guess - but that seems like overkill.

eliykat commented:
Still had to refactor tests, sorry 😁 I don't like committing redundant test code. I've refactored so that we're only testing instantiating nested objects (dates and our own nested classes).
